File: stage1/settings-plugins/src/main/java/net/rubygrapefruit/plugins/stage1/IncludedBuildPlugin.java
package net.rubygrapefruit.plugins.stage1;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.initialization.Settings;
import org.gradle.api.tasks.TaskProvider;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

@SuppressWarnings("unused")
public class IncludedBuildPlugin implements Plugin<Settings> {
    @Override
    public void apply(@NotNull Settings target) {
        target.getGradle().rootProject(project -> {
            project.getPlugins().apply("lifecycle-base");
            addLifecycle(project, "build");
            addLifecycle(project, "assemble");
            addLifecycle(project, "check");
            addTask(project, "verifySamples");
        });
    }

    private void addTask(Project project, String taskName) {
        TaskProvider<Task> provider = project.getTasks().register(taskName);
        addTaskDependencies(project, taskName, provider);
    }

    private static void addLifecycle(Project project, String taskName) {
        TaskProvider<Task> provider = project.getTasks().named(taskName);
        addTaskDependencies(project, taskName, provider);
    }

    private static void addTaskDependencies(Project project, String taskName, TaskProvider<Task> provider) {
        provider.configure(task -> task.dependsOn((Callable<List<Task>>) () -> {
            List<Task> result = new ArrayList<>();
            for (Project subproject : project.getSubprojects()) {
                Task subprojectTask = subproject.getTasks().findByName(taskName);
                if (subprojectTask != null) {
                    result.add(subprojectTask);
                }
            }
            return result;
        }));
    }
}

File: bootstrap-plugins/src/main/kotlin/net/rubygrapefruit/plugins/bootstrap/IncludedBuildPlugin.kt
package net.rubygrapefruit.plugins.bootstrap

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.api.initialization.Settings
import org.gradle.api.tasks.TaskContainer

class IncludedBuildPlugin : Plugin<Settings> {
    override fun apply(target: Settings) {
        target.run {
            gradle.rootProject { project ->
                project.run {
                    plugins.apply("lifecycle-base")
                    tasks.named("clean") {
                        it.dependsOnChildren("clean", target, project)
                    }
                    tasks.named("check") {
                        it.dependsOnChildren("check", target, project)
                    }
                    tasks.named("assemble") {
                        it.dependsOnChildren("assemble", target, project)
                    }
                    tasks.lifecycle("dist", target, project)
                    tasks.lifecycle("release", target, project)
                    tasks.lifecycle("docs", target, project)
                    tasks.lifecycle("samples", target, project)
                    tasks.lifecycle("verifySamples", target, project)
                    tasks.lifecycle("localSamples", target, project)
                }
            }
        }
    }

    private fun TaskContainer.lifecycle(name: String, target: Settings, project: Project) {
        register(name) {
            it.dependsOnChildren(name, target, project)
        }
    }

    private fun Task.dependsOnChildren(name: String, settings: Settings, rootProject: Project) {
        dependsOn(settings.gradle.includedBuilds.map { it.task(":${name}") })
        dependsOn(rootProject.subprojects.mapNotNull { it.tasks.findByName(name) })
    }
}